tao_compiler/mlir/disc/transforms/disc_llvm_insert_extract_value_simplifier.cc

```diff
@@ -0,0 +1,142 @@
+    // For safety, it only checks the insertvalue that operates on `struct`
+    // type. Meanwhile, it requires that the struct is only consumed by either
+    // insertvalue or extractvalue.
+    Type containerType = op.getContainer().getType();
+    if (!containerType.isa<LLVM::LLVMStructType>()) {
+      return failure();
+    }
+
+    DenseSet<Value> containers;
+    auto insertValueOp = op;
+    // Propagate back to find all values representing the container.
+    while (insertValueOp) {
+      auto container = insertValueOp.getContainer();
+      containers.insert(container);
+      insertValueOp = container.getDefiningOp<LLVM::InsertValueOp>();
+    }
+    // Propagate forth to find all values representing the container.
+    SmallVector<Value> forthInsertChain;
+    forthInsertChain.push_back(op);
+    while (!forthInsertChain.empty()) {
+      Value insertValueOp = forthInsertChain.back();
+      forthInsertChain.pop_back();
+      containers.insert(insertValueOp);
+      for (auto user : insertValueOp.getUsers()) {
+        if (auto insertOp = dyn_cast_or_null<LLVM::InsertValueOp>(user)) {
+          forthInsertChain.push_back(insertOp);
+        }
+      }
+    }
+
+    auto extractPosition = [](ArrayAttr attr) {
+      SmallVector<unsigned, 4> position;
+      position.reserve(attr.size());
+      for (Attribute v : attr)
+        position.push_back(v.cast<IntegerAttr>().getValue().getZExtValue());
+      return position;
+    };
+    SmallVector<unsigned, 4> position = extractPosition(op.getPosition());
+    // Check whether there are extractvalue operators on the same `position`.
+    for (auto container : containers) {
+      for (auto user : container.getUsers()) {
+        // Make sure the struct is only consumed by either insertvalue or
+        // extractvalue.
+        if (!isa<LLVM::InsertValueOp, LLVM::ExtractValueOp>(user)) {
+          return failure();
+        }
+        if (auto extract = dyn_cast_or_null<LLVM::ExtractValueOp>(user)) {
+          auto pos = extractPosition(extract.getPosition());
+          // For simplification, only check the first position.
+          if (position[0] == pos[0]) {
+            // The same first position consumerd. Current InsertValueOp cannot
+            // be erased.
+            return failure();
+          }
+        }
+      }
+    }
+
+    // Till now, this InsertValueOp is not necessary. To erase it.
+    op.replaceAllUsesWith(op.getContainer());
+
+    return success();
+  }
+};
+
+// If the extractvalue op is not consumed by other ops, erase it. It has to do
+// DCE along with LLVMInsertSimplifier to enable more simplification
+// iteratively.
+struct LLVMExtractSimplifier : public OpRewritePattern<LLVM::ExtractValueOp> {
+  explicit LLVMExtractSimplifier(MLIRContext* context)
+      : OpRewritePattern(context) {}
+  LogicalResult matchAndRewrite(LLVM::ExtractValueOp op,
+                                PatternRewriter& rewriter) const override {
+    if (op.use_empty()) {
```

Comment:
`LLVM::ExtractValueOp` does not have side effect. Can it be removed automatically if it has no other users?

Reply:
It can be eliminated automatically after this pass. But it requires to iteratively remove extractvalue and insertvalue. It cannot wait for the end of the pass.